common/src/main/scala/uk/ac/wellcome/s3/S3ObjectStore.scala

```diff
@@ -15,39 +15,55 @@ import scala.concurrent.Future
 import scala.io.Source
 import scala.util.hashing.MurmurHash3
 
-class SourcedObjectStore @Inject()(
+trait KeyPrefixGenerator[-T] {
+  def generate(obj: T): String
+}
+
+class SourcedKeyPrefixGenerator @Inject() extends KeyPrefixGenerator[Sourced] {
+  override def generate(obj: Sourced): String = {
+    val s3Shard = obj.sourceId.reverse.slice(0, 2)
+
+    s"${obj.sourceName}/${s3Shard}/${obj.sourceId}"
+  }
+}
+
+class S3ObjectStore[T] @Inject()(
   s3Client: AmazonS3,
-  @Flag("aws.s3.bucketName") bucketName: String)
-    extends Logging {
-  def put[T <: Sourced](sourcedObject: T)(
-    implicit encoder: Encoder[T]): Future[String] = {
+  @Flag("aws.s3.bucketName") bucketName: String,
+  keyPrefixGenerator: KeyPrefixGenerator[T]
+) extends Logging {
+  def put(sourcedObject: T)(implicit encoder: Encoder[T]): Future[String] = {
 
+    val keyPrefix = keyPrefixGenerator.generate(sourcedObject)
+    S3ObjectStore.put[T](s3Client, bucketName)(keyPrefix)(sourcedObject)
+  }
+
+  def get(key: String)(implicit decoder: Decoder[T]): Future[T] = {
+    S3ObjectStore.get[T](s3Client, bucketName)(key)
+  }
+}
+
+object S3ObjectStore extends Logging {
+  def put[T](s3Client: AmazonS3, bucketName: String)(keyPrefix: String)(
+    sourcedObject: T)(implicit encoder: Encoder[T]): Future[String] =
     Future.fromTry(JsonUtil.toJson(sourcedObject)).map { content =>
       val contentHash = MurmurHash3.stringHash(content, MurmurHash3.stringSeed)
 
-      // To spread objects evenly in our S3 bucket, we take the last two
-      // characters of the ID and reverse them.  This ensures that:
-      //
-      //  1.  It's easy for a person to find the S3 data corresponding to
-      //      a given source ID.
-      //  2.  Adjacent objects are stored in shards that are far apart,
-      //      e.g. b0001 and b0002 are separated by nine shards.
-      //
-      val s3Shard = sourcedObject.sourceId.reverse
-        .slice(0, 2)
+      val prefixCleaningRegex = "^/|/$".r
```

Comment:
I have no idea what this regex is meant to accomplish, or whether it's doing it correctly. Comment please!

(More thoughts: https://alexwlchan.net/2016/04/regexes-are-code/)

Reply:
Also, this feels like an odd boundary violation – both the class and the companion object have some logic for picking the key prefix. Surely better to move this up into the class, and have all the key prefix logic there?

Reply:
Okay, so Alice has explained to me that this is trimming slashes from the string. Maybe use `stripPrefix` and `stripSuffix` for readability? i.e.

```
scala> "/foo/".stripPrefix("/").stripSuffix("/")
res1: String = foo
```

Reply:
I think this should live in the object because you want to protect yourself from double slashes in both cases butt 👍 for the `stripPrefix` and `stripSuffix`. Didn't know you could do that!